File: ra-idp-go/internal/oauth2/usecases/register_client.go
// /register (RFC 7591 Dynamic Client Registration)
package usecases

import (
	"context"
	"time"

	"ra-idp-go/internal/oauth2/domain"
	"ra-idp-go/internal/oauth2/ports"
	"ra-idp-go/internal/spec"
)

type RegisterClientInput struct {
	ClientName              string
	ClientType              spec.ClientType
	RedirectURIs            []string
	GrantTypes              []spec.GrantType
	ResponseTypes           []spec.ResponseType
	TokenEndpointAuthMethod spec.TokenEndpointAuthMethod
	Scope                   string
	JWKS                    map[string]any
	JwksURI                 *string
	RequirePAR              bool
	DpopBoundAccessTokens   bool
	FapiProfile             spec.FapiProfile
}

type RegisterClientResult struct {
	Client       *spec.Client
	ClientSecret string // 平文。出力後は再表示されない (RFC 7591 §3.2.1)
}

type RegisterClientDeps struct {
	ClientRepo ports.ClientRepository
	Emit       func(spec.DomainEvent)
}

func RegisterClient(ctx context.Context, deps RegisterClientDeps, in RegisterClientInput, now time.Time) (*RegisterClientResult, error) {
	if now.IsZero() {
		now = time.Now().UTC()
	}
	if in.ClientType == "" {
		in.ClientType = spec.ClientConfidential
	}
	if len(in.RedirectURIs) == 0 {
		return nil, NewOAuthError("invalid_redirect_uri", "redirect_uris is required")
	}
	if len(in.GrantTypes) == 0 {
		in.GrantTypes = []spec.GrantType{spec.GrantAuthorizationCode}
	}
	if len(in.ResponseTypes) == 0 {
		in.ResponseTypes = []spec.ResponseType{spec.ResponseTypeCode}
	}
	if in.TokenEndpointAuthMethod == "" {
		if in.ClientType == spec.ClientPublic {
			in.TokenEndpointAuthMethod = spec.AuthMethodNone
		} else {
			in.TokenEndpointAuthMethod = spec.AuthMethodClientSecretBasic
		}
	}
	if in.TokenEndpointAuthMethod == spec.AuthMethodPrivateKeyJwt {
		candidate := spec.Client{
			ClientID: "validation", ClientType: in.ClientType,
			RedirectURIs:             []string{"https://validation.invalid/callback"},
			GrantTypes:               []spec.GrantType{spec.GrantClientCredentials},
			TokenEndpointAuthMethod:  spec.AuthMethodPrivateKeyJwt,
			JWKS:                     in.JWKS,
			JwksURI:                  in.JwksURI,
			IDTokenSignedResponseAlg: spec.SigAlgPS256,
			FapiProfile:              spec.FapiNone,
			CreatedAt:                now,
		}
		if err := candidate.Validate(); err != nil {
			return nil, NewOAuthError("invalid_client_metadata", "private_key_jwt requires non-empty inline jwks")
		}
	}
	id, err := generateOpaqueToken(16)
	if err != nil {
		return nil, err
	}
	clientID := "c_" + id
	var secret string
	var secretHash *string
	usesSecret := in.TokenEndpointAuthMethod == spec.AuthMethodClientSecretBasic ||
		in.TokenEndpointAuthMethod == spec.AuthMethodClientSecretPost
	if in.ClientType == spec.ClientConfidential && usesSecret {
		s, err := generateOpaqueToken(32)
		if err != nil {
			return nil, err
		}
		secret = s
		ss := domain.HashClientSecret(s)
		secretHash = &ss
	}
	fapiProfile := in.FapiProfile
	if fapiProfile == "" {
		fapiProfile = spec.FapiNone
	}
	scope := in.Scope
	if scope == "" {
		scope = "openid profile email"
	}
	c := &spec.Client{
		ClientID:                           clientID,
		ClientSecretHash:                   secretHash,
		ClientType:                         in.ClientType,
		RedirectURIs:                       in.RedirectURIs,
		GrantTypes:                         in.GrantTypes,
		ResponseTypes:                      in.ResponseTypes,
		TokenEndpointAuthMethod:            in.TokenEndpointAuthMethod,
		Scope:                              scope,
		JWKS:                               in.JWKS,
		JwksURI:                            in.JwksURI,
		IDTokenSignedResponseAlg:           spec.SigAlgPS256,
		RequirePushedAuthorizationRequests: in.RequirePAR,
		DpopBoundAccessTokens:              in.DpopBoundAccessTokens,
		FapiProfile:                        fapiProfile,
		CreatedAt:                          now,
	}
	if in.ClientName != "" {
		name := in.ClientName
		c.ClientName = &name
	}
	if err := c.Validate(); err != nil {
		return nil, NewOAuthError("invalid_client_metadata", err.Error())
	}
	if err := deps.ClientRepo.Save(ctx, c); err != nil {
		return nil, err
	}
	emit(deps.Emit, &spec.ClientRegistered{At: now, ClientID: clientID, ClientType: in.ClientType})
	return &RegisterClientResult{Client: c, ClientSecret: secret}, nil
}

File: ra-idp/src/oauth2/usecases/register-client.ts
/**
 * Layer 3 — Application Logic
 *
 * Dynamic Client Registration (RFC 7591) の中核。
 * 本アプリでは「管理者によるクライアント登録」のみを実装し、公開エンドポイントとしては
 * 開示しない（本番ではトークン認証付きエンドポイントとして保護する）。
 */

import { randomBytes, randomUUID, createHash } from 'crypto'
import { ClientSchema, type Client } from '../../spec-bindings/schemas'
import { OAuthError } from '../protocol/oauth-error'
import type { ClientRepository } from '../ports/client-repository'

export interface RegisterClientInput {
  client_name?: string
  client_type: 'public' | 'confidential'
  redirect_uris: string[]
  grant_types?: Client['grant_types']
  response_types?: Client['response_types']
  token_endpoint_auth_method?: Client['token_endpoint_auth_method']
  scope?: string
  /** private_key_jwt クライアントの公開鍵 (インライン JWKS)。 */
  jwks?: Record<string, unknown>
  /** private_key_jwt クライアントの JWKS エンドポイント。 */
  jwks_uri?: string
  require_pushed_authorization_requests?: boolean
  dpop_bound_access_tokens?: boolean
  fapi_profile?: Client['fapi_profile']
}

export interface RegisterClientResult {
  client: Client
  /** 機密クライアントには発行時に一度だけ平文の client_secret を返す。 */
  client_secret?: string
}

export async function registerClientUseCase(
  deps: { clientRepo: ClientRepository },
  input: RegisterClientInput,
  now: Date = new Date(),
): Promise<RegisterClientResult> {
  const client_id = `cli_${randomUUID().replace(/-/g, '')}`
  const authMethod =
    input.token_endpoint_auth_method ??
    (input.client_type === 'confidential' ? 'client_secret_basic' : 'none')

  // private_key_jwt は登録時に公開鍵を要求する (RFC 7591 §2 / ADR-008)。
  // 鍵のないクライアントは認証経路で必ず失敗するため、登録時点で弾く。
  if (authMethod === 'private_key_jwt' && !input.jwks && !input.jwks_uri) {
    throw new OAuthError(
      'invalid_request',
      'private_key_jwt クライアントは jwks または jwks_uri が必要です',
    )
  }

  // client_secret は secret ベース認証方式のときだけ発行する。
  const usesSecret = authMethod === 'client_secret_basic' || authMethod === 'client_secret_post'
  let client_secret: string | undefined
  let client_secret_hash: string | undefined
  if (input.client_type === 'confidential' && usesSecret) {
    client_secret = randomBytes(32).toString('base64url')
    client_secret_hash = createHash('sha256').update(client_secret).digest('hex')
  }
  const client = ClientSchema.parse({
    client_id,
    client_secret_hash,
    client_name: input.client_name,
    client_type: input.client_type,
    redirect_uris: input.redirect_uris,
    grant_types: input.grant_types ?? ['authorization_code', 'refresh_token'],
    response_types: input.response_types ?? ['code'],
    token_endpoint_auth_method: authMethod,
    jwks: input.jwks,
    jwks_uri: input.jwks_uri,
    scope: input.scope ?? 'openid profile email',
    require_pushed_authorization_requests: input.require_pushed_authorization_requests ?? false,
    dpop_bound_access_tokens: input.dpop_bound_access_tokens ?? false,
    fapi_profile: input.fapi_profile ?? 'none',
    created_at: now.toISOString(),
  })
  await deps.clientRepo.save(client)
  return { client, client_secret }
}
